pkg/controllers/graceful/graceful_termination_controller.go

```diff
@@ -0,0 +1,184 @@
+		Weight:          0,
+		ConnectionFlags: dst.ConnectionFlags,
+		AddressFamily:   dst.AddressFamily,
+		UpperThreshold:  dst.UpperThreshold,
+		LowerThreshold:  dst.LowerThreshold,
+	}
+	deletionTime := time.Now()
+
+	req := gracefulRequest{
+		ipvsSvc:                   svc,
+		ipvsDst:                   newDest,
+		deletionTime:              deletionTime,
+		gracefulTerminationPeriod: gracefulPeriod,
+	}
+
+	//And push it to the controller queue
+	gh.queueChan <- req
+	return nil
+}
+
+// cleanup does the lifting of removing destinations and cleaning conntrack records
+func (gh *TerminationController) cleanup() {
+	var newQueue []gracefulRequest
+	for _, dest := range gh.jobQueue {
+		// Check if our destination is old enough to consider
+		if time.Since(dest.deletionTime) > dest.gracefulTerminationPeriod {
+			glog.V(2).Infof("Deleting IPVS destination: %v", dest.ipvsDst)
+			if err := gh.ipvsHandle.DelDestination(dest.ipvsSvc, dest.ipvsDst); err != nil {
+				glog.Errorf("Failed to delete IPVS destination: %v, %s", dest.ipvsDst, err.Error())
+			}
+			// flush conntrack when endpoint for a UDP service changes
+			if dest.ipvsSvc.Protocol == syscall.IPPROTO_UDP {
+				if err := gh.flushConntrackUDP(dest); err != nil {
+					glog.Error(err.Error())
+				}
+			}
+			continue
+		}
+		newQueue = append(newQueue, dest)
+	}
+	gh.jobQueue = newQueue
+}
+
+// flushConntrackUDP flushes UDP conntrack records for the given service destination
+func (gh *TerminationController) flushConntrackUDP(dest gracefulRequest) error {
+	// Conntrack exits with non zero exit code when exiting if 0 flow entries have been deleted, use regex to check output and don't Error when matching
+	re := regexp.MustCompile("([[:space:]]0 flow entries have been deleted.)")
+	out, err := exec.Command("conntrack", "-D", "--orig-dst", dest.ipvsSvc.Address.String(), "-p", "udp", "--dport", strconv.Itoa(int(dest.ipvsSvc.Port))).CombinedOutput()
+	if err != nil {
+		if matched := re.MatchString(string(out)); !matched {
+			return fmt.Errorf("Failed to delete conntrack entry for endpoint: %s:%d due to %s", dest.ipvsSvc.Address.String(), dest.ipvsSvc.Port, err.Error())
+		}
+	}
+	glog.V(1).Infof("Deleted conntrack entry for endpoint: %s:%d", dest.ipvsSvc.Address.String(), dest.ipvsSvc.Port)
+	return nil
+}
+
+// handleReq is the function that processes incoming messages on the job queue
+func (gh *TerminationController) handleReq(req gracefulRequest) error {
+	var found bool
+	for _, dst := range gh.jobQueue {
+		if req.ipvsSvc.Address.String() == dst.ipvsSvc.Address.String() && req.ipvsSvc.Port == dst.ipvsSvc.Port && req.ipvsSvc.Protocol == dst.ipvsSvc.Protocol {
+			if req.ipvsDst.Address.String() == dst.ipvsDst.Address.String() && req.ipvsDst.Port == dst.ipvsDst.Port {
+				glog.V(2).Infof("Deletion request exists for svc: %v dst: %v", *req.ipvsSvc, *req.ipvsDst)
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		// Set the destination weight to 0
```

Comment:
I set the weight to 0 so no new connections get sent to the backend destination, while old ones are not closed because a pod is being flagged as shutting down ( good example would be rolling upgrade a webservers with graceful shutdown enabled )